From 46e4f7e06657da9380f9dc24ed63c32370576095 Mon Sep 17 00:00:00 2001
From: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Date: Fri, 10 Oct 2025 13:28:30 +0800
Subject: [PATCH] spi: aspeed: Use lower frequency to probe SPI flash

During SPI flash probe stage, lower frequency should be adopted
to avoid poor SI timing waveform quality which may be compensated
by the timing calibration mechanism after device probe stage.
Otherwise, if the higher clock frequency is used initially, the
flash ID may not be gotten and recognized.

Signed-off-by: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Change-Id: I9b65dc5e5d26dde02a29753cdb989d22374b74af
(cherry picked from commit f4ecd5cf86daa4b92d8cd79294e8b5e66cb5b46a)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/spi/spi-aspeed-smc.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/drivers/spi/spi-aspeed-smc.c b/drivers/spi/spi-aspeed-smc.c
index c31f8f62c05e..29d64a00bed2 100644
--- a/drivers/spi/spi-aspeed-smc.c
+++ b/drivers/spi/spi-aspeed-smc.c
@@ -1001,7 +1001,7 @@ static void aspeed_spi_do_calibration(struct udevice *dev)
 
 	memcpy_fromio(golden_buf, (void *)flash->ahb_base, TIMING_CAL_LEN);
 	if (!aspeed_spi_check_calib_data(golden_buf, TIMING_CAL_LEN)) {
-		dev_info(bus, "Calibration area too uniform, using low speed");
+		dev_info(bus, "Calibration area too uniform, ignore tuning\n");
 		goto no_calib;
 	}
 
@@ -1874,23 +1874,25 @@ static int aspeed_spi_set_speed(struct udevice *bus, uint hz)
 {
 	struct aspeed_spi_priv *priv = dev_get_priv(bus);
 	const struct aspeed_spi_info *info = priv->info;
+	struct aspeed_spi_plat *plat = dev_get_plat(bus);
 	struct aspeed_spi_regs *regs = priv->regs;
 	struct aspeed_spi_flash *flash;
 	enum aspeed_spi_cmd_mode mode;
 	u32 cs;
 
 	dev_dbg(bus, "%s: setting speed to %u\n", bus->name, hz);
+
 	/*
 	 * ASPEED SPI controller supports multiple CS with different
 	 * clock frequency. We cannot distinguish which CS here.
 	 * Thus, the clock frequency of all CS will be changed once.
 	 */
-
 	for (cs = 0; cs < priv->num_cs; cs++) {
 		flash = &priv->flashes[cs];
 		for (mode = CMD_USER_MODE; mode < CMD_MODE_MAX; mode++) {
 			flash->cmd_mode[mode] &= ~info->clk_ctrl_mask;
-			flash->cmd_mode[mode] |= info->get_clk_setting(bus, hz);
+			if (hz < (plat->hclk_rate / 5) || priv->disable_calib)
+				flash->cmd_mode[mode] |= info->get_clk_setting(bus, hz);
 		}
 
 		writel(flash->cmd_mode[CMD_READ_MODE], &regs->ce_ctrl[cs]);
-- 
2.34.1

